### DISCUSSION TITLE
Python script for generating custom flavors for megalinter

### DISCUSSION THREAD

**Discussion by Heyzi** [oxsecurity/megalinter#3972] [Show and tell]
For my current project, I developed a Python script that enables the creation of a Docker image containing only selected linters. This tool should prove useful for many, as not everyone requires the extensive set of linters typically included. 

The script fulfills my main task, I don't plan to significantly improve it further.

You can find it in my GitHub repository: https://github.com/Heyzi/megalinter_flavor_generator

**Comment by llaville** [comment]
@Heyzi Idea is cool. I'll have a look later on this, but I love this feature !

**Comment by llaville** [comment]
@Heyzi When we can read `python script_name.py` on your project page, I suppose we should see `python flavor_generator.py`; right ?

**Reply by Heyzi** [reply]
Yeah, that's right.
I'll fix the readme for clarity.

**Comment by Heyzi** [comment]
The logic of the script is as follows:

```
1. Update the schema file (megalinter-descriptor.jsonschema.json) to include the new flavor if it doesn't exist.
2. Modify the flavor factory file (flavor_factory.py) to add the new flavor and its description.
3. Update YAML descriptor files in the descriptors directory:

- Add the new flavor to linters specified in the components list.
- Remove the new flavor from linters not in the components list.

4. Run the build.py script to generate necessary files for the new flavor.
5. Build a Docker image for the new flavor using the appropriate Dockerfile.
```

**Comment by Heyzi** [comment]
I've also created a lightweight version of the script called `flavor_generator_light.py`. This version reuses some functions from `build.py` to generate only the essential files: `action.yml`, `Dockerfile`, and `flavor.json`.

While this approach isn't ideal for production use 🙃, it can be beneficial for users who need quick results. With this lightweight script, users can generate the necessary Docker image much faster...

**Comment by nvuillam** [comment]
I like this "BYOM" (Bring Your Own MegaLinter) pattern :)

We probably could industrialize that even more in the future, some tool that would:
- Take just a list of linter identifiers as input parameters
- Download latest sources of MegaLinter
- Build the custom Dockerfile ( = your script ^^)
- Build the generated docker image from Dockerfile
- Upload it on a Docker / ghcr.io account :)

It could even be a GitHub Actions :)

**Comment by nwiltsie** [comment]
I really like this - I forked it (https://github.com/nwiltsie/megalinter_flavor_generator) and extended it roughly the way @nvuillam laid out.

The key additions are:

* Megalinter is a submodule, kept up-to-date by Dependabot
* On every push to `main`, a workflow re-generates the custom flavor's Dockerfile, builds _only_ that image, and pushes it to ghcr.io.

Springboarding off of @Heyzi's awesome work, that means that I have a custom flavor, defined by the components that I want, kept up-to-date with Megalinter core: https://github.com/nwiltsie/megalinter_flavor_generator/pkgs/container/bl-megalinter

**Reply by nvuillam** [reply]
That's amazing :)

If you want to really mirror version management, you can do something like:

- when new push to MegaLinter main -> generate beta version of your flavor
- when new GitHub release -> generate same tag version :)
